compiler/rustc_const_eval/src/interpret/place.rs

```diff
@@ -869,8 +870,20 @@ where
             Ok(src_val) => {
                 assert!(!src.layout.is_unsized(), "cannot have unsized immediates");
                 // Yay, we got a value that we can write directly.
-                // FIXME: Add a check to make sure that if `src` is indirect,
-                // it does not overlap with `dest`.
+                // Then make sure `src` does not overlap with `dest`.
+                if let Operand::Indirect(src_mplace) = src.deref()
+                    && let Place::Ptr(dest_mplace) = dest.deref()
+                    && let Ok((src_alloc_id, src_offset, _)) = self.ptr_try_get_alloc_id(src_mplace.ptr)
+                    && let Ok((dest_alloc_id, dest_offset, _)) = self.ptr_try_get_alloc_id(dest_mplace.ptr)
+                    && src_alloc_id == dest_alloc_id
+                {
+                    let size = src_val.layout.size;
+                    if (src_offset <= dest_offset && src_offset + size > dest_offset)
```

Comment:
please pull these range checks out into a closure, they seem like a copy paste hazard ^^

Reply:
Thanks for pointing out. I made some change to use a `check_ptr_overlap` nethod here. And a small opt in `mem_copy_repeatedly`: if `src` doesn't overlap with `dest`, we can always use `ptr::copy_nonoverlapping`